tools/bpf2c/bpf_code_generator.cpp

```diff
@@ -707,14 +723,94 @@ bpf_code_generator::encode_instructions(const bpf_code_generator::unsafe_string&
                 break;
             case EBPF_SIZE_W:
                 size_type = "uint32_t";
+                lock_type = "volatile long";
                 break;
             case EBPF_SIZE_DW:
+                size_num = "64";
                 size_type = "uint64_t";
+                lock_type = "volatile int64_t";
                 break;
             }
+            raw_source = source;
             source = "(" + size_type + ")" + source;
-            output.lines.push_back(
-                std::format("*({}*)(uintptr_t)({} + {}) = {};", size_type, destination, offset, source));
+            if ((inst.opcode & EBPF_MODE_MASK) == EBPF_MODE_ATOMIC) { // MODE_ATOMIC
+                auto line = std::string("");
+                switch (inst.imm) {
+                case EBPF_ATOMIC_ADD:
+                case EBPF_ATOMIC_ADD | EBPF_ATOMIC_FETCH:
+                    line = std::format(
+                        "_InterlockedExchangeAdd{}(({}*)(uintptr_t)({} + {}), {});",
```

Comment:
Is there a reason why we could not use the windows version of [InterlockedAdd](https://learn.microsoft.com/en-us/windows/win32/api/winnt/nf-winnt-interlockedadd) (and for other interlocked intrinsics used in this PR) instead? Windows has equivalents for all the clang ```_InterLocked...``` intrinsics used here.

That might also do away with the necessity of having to include clang's ```intrin.h```

Reply:
`instrin.h` is also in MSVC (C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.33.31629\include\intrin.h). As I know, we use cl.exe to build generated code in this project rather than clang, so using `_InterLocked*` here has no dependency on clang.

Using `_Interlocked*` or `Interlocked*` is both ok, one include `intrin.h`, another include `winnt.h`.

Reply:
AFAIK we prefer the ```Interlocked...``` versions from ```winnt.h``` at least insofaras windows specific components are concerned.  That being said, I'm relatively new to this project as well, so I'll defer to the veterans on this one :-)

@Alan-Jowett, @dthaler: Thoughts?

Reply:
Any update? @dv-msft

Reply:
Could you please add the tests recommended by @Alan-Jowett in his comment? That emitted C code output might give us better picture.